### PR TITLE
gh-78469: Declare missing sethostname for Solaris 10

### PR DESCRIPTION
**Pull request by kulikjak**
As discussed in the issue, `sethostname` is missing from Solaris 10 headers but is available on Solaris 11.

And because there isn't an OS version specific macro for Solaris, I had to "create" one.

* Issue: gh-78469

### CODE REVIEW

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -5652,8 +5652,9 @@ socket_sethostname(PyObject *self, PyObject *args)
     Py_buffer buf;
     int res, flag = 0;
 
-#ifdef _AIX
-/* issue #18259, not declared in any useful header file */
+#if defined(_AIX) || (defined(__sun) && defined(__SVR4) && SUNOS_VERSION == 510)
```
Comment (vstinner) [inline Modules/socketmodule.c]: wow, that's very spefic. It was available *before* and *after*, but not at version 510? :-)

File: configure.ac [hunk 2]
```diff
@@ -590,6 +590,14 @@ then
 	darwin*) MACHDEP="darwin";;
 	'')	MACHDEP="unknown";;
     esac
+
+    if test "$ac_sys_system" = "SunOS"; then
+	# For Solaris, there isn't an OS version specific macro defined
+	# in most compilers, so we define one here.
+	SUNOS_VERSION=`echo $ac_sys_release | tr -d '.'`
+	AC_DEFINE_UNQUOTED([SUNOS_VERSION], [$SUNOS_VERSION],
+	                   [The version of SunOS/Solaris as reported by `uname -r' without the dot.])
```
Comment (vstinner) [inline configure.ac:599]: How do you distinguish Solaris 5.10 and Solaris 51.0?

Maybe multiply the major version by 100 or 1000? or use a shift of 8 bits? The usage would be:

```
major = VERSION >> 8
minor = VERSION & 255
```

The disavantage is the need to write versions in hexadecimal :-(

Comment (vstinner) [reply configure.ac:599]: @erlend-aasland: Should we prefix new pyconfig.h constants by Py_ ? see: https://github.com/capi-workgroup/problems/issues/46

**Comment by vstinner** [comment]
When was Solaris 10 release? Is it still supported? Does this change better to forks of Solaris?

File: configure.ac (hunk 2 above)
Comment (erlend-aasland) [reply configure.ac:599]: I think we should follow the PEP-7 recommendations: https://peps.python.org/pep-0007/#naming-conventions

Comment (kulikjak) [reply configure.ac:599]: > How do you distinguish Solaris 5.10 and Solaris 51.0?

I see what you mean, but that's not happening. The leading 5 was dropped from the name 25 years ago and while e.g. `uname` and compilation targets still use it, nobody today calls it Solaris/SunOS 5.10; only Solaris 10. If we ever release Solaris 51, `uname -r` should report it as 5.51.

That said, I am not strongly against changing it if you still prefer your suggestion.

> I think we should follow the [PEP-7](https://peps.python.org/7) recommendations: https://peps.python.org/pep-0007/#naming-conventions

Thanks, I missed that. It's fixed now.

File: configure.ac [hunk 3]
```diff
@@ -590,6 +590,14 @@ then
 	darwin*) MACHDEP="darwin";;
 	'')	MACHDEP="unknown";;
     esac
+
+    if test "$ac_sys_system" = "SunOS"; then
+	# For Solaris, there isn't an OS version specific macro defined
+	# in most compilers, so we define one here.
+	SUNOS_VERSION=`echo $ac_sys_release | tr -d '.'`
```
Comment (vstinner) [inline configure.ac]: I suggest:

```suggestion
	SUNOS_VERSION=`echo $ac_sys_release | sed -e 's!\.\([0-9]\)$!.0\1!g' | tr -d '.'`
```

I'm not sure about how you build the number:

```
vstinner@mona$ echo 5.1 | tr -d '.'
51
vstinner@mona$ echo 5.10 | tr -d '.'
510
```

I suggest to make sure that you have 2 digits after the dot before removing the dot:

```
vstinner@mona$ echo 5.1 | sed -e 's!\.\([0-9]\)$!.0\1!g' | tr -d '.'
501
vstinner@mona$ echo 5.10 | sed -e 's!\.\([0-9]\)$!.0\1!g' | tr -d '.'
510
```

So Oracle Solaris 5.9 (released in 2002) is formatted as the number 509.

In Python, we had many issues when we switched from Python 3.9 to 3.10, a lot of code was hardcoded to parse the version as 3 characters or compared versions as floatting point number:

```
>>> 3.1 < 3.10
False
```

Comment (kulikjak) [reply configure.ac]: That I didn't consider, that's a good point. It's updated in both files.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Is this a bug-fix? If so, we should backport it and add a NEWS entry.

**Comment by vstinner** [comment]
> @kulikjak Update configure 

Do you use "make regen-configure"? See: https://docs.python.org/dev/using/configure.html#generated-files

@kulikjak: Anyway, I push directly into your branch to fix the configure script (regenerate it).

File: Modules/socketmodule.c [hunk 4]
```diff
@@ -5652,8 +5652,9 @@ socket_sethostname(PyObject *self, PyObject *args)
     Py_buffer buf;
     int res, flag = 0;
 
-#ifdef _AIX
-/* issue #18259, not declared in any useful header file */
+#if defined(_AIX) || (defined(__sun) && defined(__SVR4) && Py_SUNOS_VERSION == 510)
```
Comment (vstinner) [inline Modules/socketmodule.c]: I expected that older versions would also need it:

```suggestion
#if defined(_AIX) || (defined(__sun) && defined(__SVR4) && Py_SUNOS_VERSION <= 510)
```

**Review by vstinner — APPROVED** [review]
LGTM.

**Comment by vstinner** [comment]
@kulikjak: I also replaced `Py_SUNOS_VERSION == 510` with `Py_SUNOS_VERSION <= 510`.

**Comment by kulikjak** [comment]
> Do you use "make regen-configure"? See: https://docs.python.org/dev/using/configure.html#generated-files
Sigh, sorry. Normally, I do, but this time I thought that I could quickly edit it myself, ... and I was wrong. Thanks for fixing it.

> Is this a bug-fix? If so, we should backport it and add a NEWS entry.
I guess it is (fixing a build on Solaris <= 10). I can add a NEWS entry, although now there is a `skip-news` tag, so I am not sure whether I should do so?

Other than that, the build failure doesn't seem to be related, so that's good.

**Review by Dezzj — COMMENTED** [review]
Veiwed

**Pull request merged**